ext/openssl/ossl_mac.c

```diff
@@ -0,0 +1,174 @@
+#include "ossl.h"
+
+#if OSSL_OPENSSL_PREREQ(3, 0, 0)
+
+#define NewMAC(klass)                                   \
+    TypedData_Wrap_Struct((klass), &ossl_mac_type, 0)
+#define SetMAC(obj, ctx) do {                                           \
+        if (!(ctx))                                                     \
+            ossl_raise(rb_eRuntimeError, "MAC wasn't initialized");     \
+        RTYPEDDATA_DATA(obj) = (ctx);                                   \
+    } while (0)
+#define GetMAC(obj, ctx) do {                                           \
+        TypedData_Get_Struct((obj), EVP_MAC_CTX, &ossl_mac_type, (ctx)); \
+        if (!(ctx))                                                     \
+            ossl_raise(rb_eRuntimeError, "MAC wasn't initialized");     \
+    } while (0)
+
+/*
+ * Classes
+ */
+static VALUE cMAC;
+static VALUE cCMAC;
+static VALUE eMACError;
+
+/*
+ * Public
+ */
+
+/*
+ * Private
+ */
+static void
+ossl_mac_free(void *ctx)
+{
+    EVP_MAC_CTX_free(ctx);
+}
+
+static const rb_data_type_t ossl_mac_type = {
+    "OpenSSL/MAC",
+    {
+        0, ossl_mac_free,
+    },
+    0, 0, RUBY_TYPED_FREE_IMMEDIATELY | RUBY_TYPED_WB_PROTECTED,
+};
+
+static VALUE
+ossl_mac_alloc(VALUE klass)
+{
+    return NewMAC(klass);
+}
+
+static VALUE
+ossl_mac_initialize(VALUE self, VALUE algorithm)
+{
+    EVP_MAC *mac;
+    EVP_MAC_CTX *ctx;
+
+    mac = EVP_MAC_fetch(NULL, StringValueCStr(algorithm), NULL);
+    if (!mac)
+        ossl_raise(eMACError, "EVP_MAC_fetch");
+    ctx = EVP_MAC_CTX_new(mac);
+    if (!ctx) {
+        EVP_MAC_free(mac);
+        ossl_raise(eMACError, "EVP_MAC_CTX_new");
+    }
+    SetMAC(self, ctx);
```

Comment:
Technically, `#initialize` and `#initialize_copy` can be called twice, and in that case `EVP_MAC_CTX` already stored in `RTYPEDDATA_DATA(self)` will leak. It should be prevented.

I suggest taking a look at `ossl_pkey*.c` which deals with this issue by raising an exception.